### PR TITLE
[TRTLLM-8414][chore] BREAKING CHANGE: refine sampling strategy selection

### PR DESCRIPTION
**Pull request by ixlmar**
## Description

**NOTE:** These changes specify the sampling behavior in some previously undefined cases. While the new behavior should be in line with the examples in the documentation, it does differ from the base-branch behavior in certain cases.

Roughly, the new semantics are (see TRTLLM-8414 for complete background):
* If `top_k = top_p = temperature = None`  (the default if the respective parameters are not specified), sampling is greedy (`torch.argmax`).
* If a non-trivial `top_k`, `top_p`, and/or `temperature` is provided, sampling proceeds accordingly. In this case, unspecified parameters default to `top_k = 0`,  `top_p = 1`, `temperature = 1.0`.
* If either `temperature = 0`, `top_p = 0`,  and/or `top_k = 1`, is specified, sampling is greedy, irrespective of the values of the remaining parameters.

## Test Coverage

Unit tests for the sampling strategy selection are implemented as part of this PR.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Summary by CodeRabbit

- New Features
  - Unified sampling strategies with support for top-k, top-p (nucleus), temperature, and their combinations, with clearer greedy behavior when parameters imply determinism.
- Bug Fixes
  - Improved handling of edge cases and defaults in sampling (e.g., validation of temperature/top-p/top-k ranges and greedy fallbacks).
- Tests
  - Added comprehensive unit tests covering strategy selection, parameter validation, and draft-probability decisions.
  - Enabled selection of a specific sampler type in tests.
  - Suppressed noisy traces for skipped tests.
- Chores
  - Adjusted tooling configuration to exclude a specific test file from formatting and spell-checking.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 1]
```diff
@@ -1706,8 +1770,17 @@ def _process_requests(
...
+        top_p = params.top_p
+        top_k = params.top_k
+
         # Do not request draft probs when sampling is greedy.
-        return _request_strategy(request) is not GREEDY
+        return SamplingParams.params_imply_greedy_decoding(
+            temperature=temperature,
+            top_p=top_p,
+            top_k=top_k,
+        )
```
[the remark this answers is not in the document]
Comment (ixlmar) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:1783]: Good catch

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 2]
```diff
@@ -195,106 +196,104 @@ def is_generation_model(self) -> bool:
...
-    temperature: float = 1.0,
+    top_k: int,
+    top_p: float,
+    temperature: float,
     generator: Optional[torch.Generator] = None
 ) -> tuple[torch.Tensor, torch.Tensor]:
     logits_dim = logits.dim()
     assert logits_dim == 2, "logits should be 2D: [batch_size, vocab_size]"
+    assert temperature > 0, "non-greedy sampling requires valid temperature"
+    logits = logits / max(temperature, 1e-5)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:258]: Why do we need the `max` if we check that `temperature > 0` before?

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 3]
```diff
@@ -195,106 +196,104 @@ def is_generation_model(self) -> bool:
...
-
-    # set the logits to -inf whose is outside top_p
-    indices_to_remove = sorted_indices_to_remove.scatter(
-        1, sorted_indices, sorted_indices_to_remove)
-    logits = logits.masked_fill(indices_to_remove, float('-inf'))
+    # top-p: mask out logits outside the nucleus
+    if need_top_p:
+        sorted_logits, sorted_indices = torch.sort(logits,
+                                                   descending=True,
+                                                   dim=-1)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:279]: Use stable sort? (see coderabbit nitpick comment)

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 4]
```diff
@@ -195,106 +196,104 @@ def is_generation_model(self) -> bool:
...
+        #     Thus, top_p -> 0 resembles greedy; agreement requires torch.sort(..., stable=True).
+        sorted_indices_to_remove = cumulative_probs > top_p
+        sorted_indices_to_remove[:,
+                                 1:] = sorted_indices_to_remove[:, :-1].clone()
+        sorted_indices_to_remove[:, 0] = 0
+
+        # set the logits to -inf for token indices outside top_p
+        indices_to_remove = sorted_indices_to_remove.scatter(
+            1, sorted_indices, sorted_indices_to_remove)
+        logits = logits.masked_fill(indices_to_remove, float('-inf'))
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:296]: Is there a reason why we use `torch.where` for `top-k` but `masked_fill` for `top-p`?

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 5]
```diff
@@ -359,48 +358,100 @@ def sample_rejected(draft_probs: torch.Tensor, target_probs: torch.Tensor,
...
+        cast(Optional[List[float]], sampling_config.temperature))
+    top_p = _unwrap_singleton(cast(Optional[List[float]],
+                                   sampling_config.top_p))
+    top_k = _unwrap_singleton(cast(Optional[List[int]], sampling_config.top_k))
+
+    return TorchSamplerSamplingParams(
+        temperature=temperature,
+        top_p=top_p,
+        top_k=top_k,
+    )
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:403]: I think this file got a bit bloated with all these helper functions. Do you think we should move them to a separate file?

File: tests/integration/test_lists/test-db/l0_a10.yml [hunk 6]
```diff
@@ -14,6 +14,7 @@ l0_a10:
       backend: pytorch
   tests:
   # ------------- PyTorch tests ---------------
+  - unittest/_torch/sampler/test_torch_sampler.py
```
Comment (Funatiq) [inline tests/integration/test_lists/test-db/l0_a10.yml:17]: FYI: All sampler tests also run on H100 and B200 if they are not marked with `@force_ampere` or something similar.

File: pyproject.toml [hunk 7]
```diff
@@ -144,6 +146,7 @@ include = [
     "tests/unittest/_torch/modeling/test_modeling_mistral.py",
     "tests/unittest/_torch/modeling/test_modeling_pixtral.py",
     "tests/unittest/_torch/models/checkpoints/hf/test_weight_loader.py",
+    "tests/unittest/_torch/sampler/test_torch_sampler.py",
```
Comment (Funatiq) [inline pyproject.toml:149]: Do you think we could make ruff the default for new files instead?

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 4 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:296]: Good question. Optimizing the actual ops was out of scope for this PR, but certainly this is something I will keep in mind for follow-up changes.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 5 above)
Comment (ixlmar) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:403]: Yes, let's do that. I will try to include it into the next PR.

**Review by Funatiq — APPROVED** [review]

**Review by lucaslie — APPROVED** [review]
lgtm
